core/string/optimized_translation.cpp

```diff
@@ -52,62 +52,74 @@ void OptimizedTranslation::generate(const Ref<Translation> &p_from) {
 
 	int size = Math::larger_prime(keys.size());
 
-	Vector<Vector<Pair<int, CharString>>> buckets;
+	struct TrItem {
+		uint32_t idx = 0;
+		CharString cs;
+		uint32_t plural = 0;
+	};
+
+	Vector<Vector<TrItem>> buckets;
 	Vector<HashMap<uint32_t, int>> table;
 	Vector<uint32_t> hfunc_table;
 	Vector<CompressedString> compressed;
 
 	table.resize(size);
 	hfunc_table.resize(size);
 	buckets.resize(size);
-	compressed.resize(keys.size());
 
 	int idx = 0;
 	int total_compression_size = 0;
 
+	set_plural_rule(p_from->get_plural_rule());
+
 	for (const StringName &E : keys) {
-		//hash string
-		CharString cs = E.operator String().utf8();
-		uint32_t h = hash(0, cs.get_data());
-		Pair<int, CharString> p;
-		p.first = idx;
-		p.second = cs;
-		buckets.write[h % size].push_back(p);
-
-		//compress string
-		CharString src_s = p_from->get_message(E).operator String().utf8();
-		CompressedString ps;
-		ps.orig_len = src_s.size();
-		ps.offset = total_compression_size;
-
-		if (ps.orig_len != 0) {
-			CharString dst_s;
-			dst_s.resize(src_s.size());
-			int ret = smaz_compress(src_s.get_data(), src_s.size(), dst_s.ptrw(), src_s.size());
-			if (ret >= src_s.size()) {
-				//if compressed is larger than original, just use original
-				ps.orig_len = src_s.size();
-				ps.compressed = src_s;
+		Vector<String> srcs = p_from->get_plural_messages(E);
+		for (int n = 0; n < srcs.size(); n++) {
+			//hash string
```

Comment:
```suggestion
			// Hash string.
```
Same with other comments.